design/codeset.go

```diff
@@ -0,0 +1,141 @@
+		HTTP(func() {
+			// Requests to the service consist of HTTP GET requests.
+			// The payload fields are encoded as path parameters.
+			GET("/codesets")
+			Param("project", String, "List only Codesets that belong to given project", func() {
+				Example("mlflow-project-01")
+			})
+			Param("label", String, "List only Codesets with matching label", func() {
+				Example("mlflow")
+			})
+			// Responses use a "200 OK" HTTP status.
+			// The result is encoded in the response body (default).
+			Response(StatusOK)
+			Response("NotFound", StatusNotFound)
+		})
+
+		// GRPC describes the gRPC transport mapping.
+		GRPC(func() {
+			// Responses use a "OK" gRPC code.
+			// The result is encoded in the response message (default).
+			Response(CodeOK)
+			Response("NotFound", CodeNotFound)
+		})
+
+	})
+
+	Method("register", func() {
+		Description("Register a Codeset with the FuseML codeset store.")
+
+		// Payload also accepts a Type object where you can list its attribute
+		// as well as its required fields
+		Payload(Codeset, "Codeset descriptor")
+
+		Error("BadRequest", func() {
+			Description("If the Codeset does not have the required fields, should return 400 Bad Request.")
+		})
+
+		Result(Codeset)
+
+		HTTP(func() {
+			POST("/codesets")
+			Response(StatusCreated)
+			Response("BadRequest", StatusBadRequest)
+		})
+
+		GRPC(func() {
+			Response(CodeOK)
+			Response("BadRequest", CodeInvalidArgument)
+		})
+	})
+
+	Method("get", func() {
+		Description("Retrieve an Codeset from FuseML.")
+
+		Payload(func() {
+			Field(1, "project", String, "Project name", func() {
+				Example("mlflow-project-01")
+			})
+			Field(2, "name", String, "Codeset name", func() {
+				Example("mlflow-app-01")
+			})
+			Required("project", "name")
+		})
+
+		Error("BadRequest", func() {
+			Description("If neither name or project is not given, should return 400 Bad Request.")
+		})
+		Error("NotFound", func() {
+			Description("If there is no codeset with the given name and project, should return 404 Not Found.")
+		})
+
+		Result(Codeset)
+
+		HTTP(func() {
+			GET("/codesets/{project}/{name}")
+			Response(StatusOK)
+			Response("BadRequest", StatusBadRequest)
+			Response("NotFound", StatusNotFound)
+		})
+
+		GRPC(func() {
+			Response(CodeOK)
+			Response("BadRequest", CodeInvalidArgument)
+			Response("NotFound", CodeNotFound)
+		})
+	})
+})
+
+var Codeset = Type("Codeset", func() {
+
+	Field(1, "name", String, "The name of the Codeset", func() {
+		Example("mlflow-app-01")
+	})
+	Field(2, "project", String, "The project this Codeset belongs to", func() {
+		Example("mlflow-project-01")
+	})
+	Field(3, "location", String, "Path to the code that should be registered as Codeset", func() {
+		Example("work/ml/mlflow-code")
+	})
+	Field(4, "description", String, "Codeset description", func() {
+		Example("My first MLFlow application with FuseML")
+	})
+	Field(5, "label", String, "Additional Codeset label that helps with identifying the type", func() {
```

Comment:
We should support associating not one but several labels with a single codeset. It would be even better if labels were modeled as dictionaries instead of lists of strings, because the dict syntax is better suited for matching purposes. For example, keys can represent classes of capabilities (e.g. library, app, predictor) that users and pipelines can filter and match to the values they support (e.g. `app: mlflow` and `library: sklearn` can be used to filter codesets that represent MLFlow projects using the scikit-learn library).

Reply:
Yes, I know this should be a list, I've just started simple

Reply:
I change the labels type to list of strings